pkg/clusteragent/orchestrator/controller.go

```diff
@@ -0,0 +1,126 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build kubeapiserver,orchestrator
+
+package orchestrator
+
+import (
+	"math/rand"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	processcfg "github.com/DataDog/datadog-agent/pkg/process/config"
+	"github.com/DataDog/datadog-agent/pkg/process/util/api"
+	"github.com/DataDog/datadog-agent/pkg/process/util/orchestrator"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ControllerContext holds necessary context for the controller
+type ControllerContext struct {
+	UnassignedPodInformerFactory informers.SharedInformerFactory
+	Client                       kubernetes.Interface
+	StopCh                       chan struct{}
+	Hostname                     string
+	ClusterName                  string
+	ConfigPath                   string
+}
+
+// Controller is responsible of collecting & sending orchestrator info
+type Controller struct {
+	unassignedPodLister     corelisters.PodLister
+	unassignedPodListerSync cache.InformerSynced
+	groupID                 int32
+	hostName                string
+	clusterName             string
+	apiClient               api.Client
+	processConfig           *processcfg.AgentConfig
+}
+
+// StartController starts the orchestrator controller
+func StartController(ctx ControllerContext) error {
+	if !config.Datadog.GetBool("orchestrator_explorer.enabled") {
+		log.Info("orchestrator explorer is disabled")
+		return nil
+	}
+	if ctx.ClusterName == "" {
+		log.Info("orchestrator explorer enabled but no cluster name set: disabling")
+		return nil
+	}
+	orchestratorControler := newController(ctx)
+
+	go orchestratorControler.Run(ctx.StopCh)
+
+	ctx.UnassignedPodInformerFactory.Start(ctx.StopCh)
+
+	return nil
+}
+
+func newController(ctx ControllerContext) *Controller {
+	podInformer := ctx.UnassignedPodInformerFactory.Core().V1().Pods()
+	oc := &Controller{
+		unassignedPodLister:     podInformer.Lister(),
+		unassignedPodListerSync: podInformer.Informer().HasSynced,
+		groupID:                 rand.Int31(),
```

Comment:
Out of curiosity - how is this used?

Reply:
It’s the batch identifier - it’s mainly useful when we need to track batches on the backend. 

Examples of things we’ve done in various places:
- Determine when a previous batch has finished being sent as we’ve seen all N payloads for Batch X.
- Skip data for every other batch for a host for some given processing workload.